### PR TITLE
enable find in cell output when output is a data stream

### PR DESCRIPTION
**Pull request by MaddyDev**
This PR is an initial step to enable find in cell output. Changes here enable find when the cell output is a stream. 

As part of this, I refactored the addDecorations/removeDecorations code from text cell component to its abstract parent class 'CellView' so that it can be reused by output component.

findInCellOutput

findSQLResultSets

### CODE REVIEW

**Comment by lucyzhang929** [comment]
Hey Maddy, while testing I noticed that when I search for "hello", "print" also gets highlighted in the cell below:
Screen Shot 2021-11-30 at 7 47 03 AM

This happens for new untitled notebooks as well as if I add this cell to an existing notebook, "print" always gets highlighted when searching for "hello".

**Comment by MaddyDev** [comment]
> Hey Maddy, while testing I noticed that when I search for "hello", "print" also gets highlighted in the cell below: Screen Shot 2021-11-30 at 7 47 03 AM
> 
> This happens for new untitled notebooks as well as if I add this cell to an existing notebook, "print" always gets highlighted when searching for "hello".

I can repro it too, I guess I regressed something during clean up, thanks for catching this, will look into it.

**Comment by chlafreniere** [comment]
tests? 😄

File: src/sql/workbench/contrib/notebook/browser/notebookStyles.ts [hunk 1]
```diff
@@ -250,7 +250,7 @@ export function registerNotebookThemes(overrideEditorThemeSetting: boolean, conf
 			collector.addRule(`.notebook-preview .rangeHighlight { background-color: ${notebookFindMatchHighlightColor};}`);
 		}
 		if (notebookFindRangeHighlightColor) {
-			collector.addRule(`.notebook-preview .rangeSpecificHighlight { background-color: ${notebookFindRangeHighlightColor}!important;}`);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookStyles.ts:253]: Why are you changing this? And what about the notebookFindMatchHighlightColor above?

File: src/sql/workbench/contrib/notebook/browser/cellViews/interfaces.ts [hunk 2]
```diff
@@ -3,15 +3,29 @@
...
+import { ICellModel } from 'sql/workbench/services/notebook/browser/models/modelInterfaces';
 
+const findHighlightClass = 'rangeHighlight';
+const findRangeSpecificClass = 'rangeSpecificHighlight';
 export abstract class CellView extends AngularDisposable implements OnDestroy, ICellEditorProvider {
+
+	protected _isFindActive: boolean = false;
+	protected _highlightRange: NotebookRange;
+	protected output: ElementRef;
+	protected _notebookService: INotebookService;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/cellViews/interfaces.ts:25]: Typically only private vars should start with _

File: src/sql/workbench/services/notebook/browser/notebookService.ts [hunk 3]
```diff
@@ -199,11 +200,13 @@ export class NotebookRange extends Range {
 	}
 	cell: ICellModel;
 	isMarkdownSourceCell: boolean;
+	isCodeOutput: boolean;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/notebookService.ts:203]: If this is going to be generic across different types of cells and whatnot then consider just having a single sourceType or something to cut down on the amount of `is...` vars you have to add (you already have two now with this new one)

File: src/sql/workbench/contrib/notebook/browser/cellViews/interfaces.ts [hunk 4]
```diff
@@ -29,8 +44,122 @@ export abstract class CellView extends AngularDisposable implements OnDestroy, I
...
+
+
+	protected getChildren(parent: any): any[] {
+		let children: any = [];
+		if (parent.children.length > 1 && parent.nodeName.toLowerCase() !== 'li' && parent.nodeName.toLowerCase() !== 'p') {
+			for (let child of parent.children) {
+				children = children.concat(this.getChildren(child));
+			}
+		} else {
+			return parent;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/cellViews/interfaces.ts:158]: You're returning a single object here, but the return type is an array.

File: src/sql/workbench/contrib/notebook/browser/cellViews/interfaces.ts [hunk 5]
```diff
@@ -29,8 +44,122 @@ export abstract class CellView extends AngularDisposable implements OnDestroy, I
...
+				} else {
+					children.push(element);
+				}
+			}
+		}
+		return children;
 	}
+
+
+	protected getChildren(parent: any): any[] {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/cellViews/interfaces.ts:151]: These should be typed.

Also the name `getChildren` is very generic for what seems to be a specialized form of getting the children. Please at least add a doc comment, and ideally find a better name for the function

(I'd also suggest moving this out of the class - it has no reason to be here. Either put it in a generic utils type module or you can keep it in this module if you think this isn't going to be used anywhere else)

File: src/sql/workbench/contrib/notebook/browser/cellViews/output.component.ts [hunk 6]
```diff
@@ -187,4 +192,116 @@ export class OutputComponent extends CellView implements OnInit, AfterViewInit {
...
+				grids?.forEach(g => {
+					markAllOccurances = new Mark(g);
+					markAllOccurances.unmark({ acrossElements: true, className: findHighlightClass });
+					markAllOccurances.unmark({ acrossElements: true, className: findRangeSpecificClass });
+				});
+			}
+		}
+	}
+
+	protected override getHtmlElements(): any[] {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/cellViews/output.component.ts:288]: Properly type the return type

File: src/sql/workbench/contrib/notebook/browser/notebookStyles.ts (hunk 1 above)
Comment (MaddyDev) [reply src/sql/workbench/contrib/notebook/browser/notebookStyles.ts:253]: Previously we only looked at rendered markdown which is under .notebook-preview class but now to extend it to outputs which does not fall under that, I removed it and made more generic.

File: src/sql/workbench/contrib/notebook/browser/cellViews/interfaces.ts (hunk 5 above)
Comment (MaddyDev) [reply src/sql/workbench/contrib/notebook/browser/cellViews/interfaces.ts:151]: Angular doesn't specify a type for the native html element of the ElementRef, that is why I followed what angular has which is 'any'.  And this is very specific to find and I don't see it being reused anywhere. How about getHtmlNodeChildren for getChildren?

File: src/sql/workbench/contrib/notebook/browser/cellViews/codeCell.component.ts [hunk 7]
```diff
@@ -82,8 +81,10 @@ export class CodeCellComponent extends CellView implements OnInit, OnChanges {
 		if (this.codeCells) {
 			editors.push(...this.codeCells.toArray());
 		}
-		if (this.outputCells) {
-			editors.push(...this.outputCells.toArray());
+		if (this.outputAreaCell) {
+			if (this.outputAreaCell) {
```
Comment (barbaravaldez) [inline src/sql/workbench/contrib/notebook/browser/cellViews/codeCell.component.ts:85]: This should only be one condition

File: src/sql/workbench/contrib/notebook/browser/cellViews/output.component.ts [hunk 8]
```diff
@@ -6,6 +6,8 @@ import 'vs/css!./code';
 import 'vs/css!./media/output';
 
 import { OnInit, Component, Input, Inject, ElementRef, ViewChild, SimpleChange, AfterViewInit, forwardRef, ChangeDetectorRef, ComponentRef, ComponentFactoryResolver } from '@angular/core';
+import * as Mark from 'mark.js';
+
```
Comment (barbaravaldez) [inline src/sql/workbench/contrib/notebook/browser/cellViews/output.component.ts:10]: nit: extra line

File: src/sql/workbench/contrib/notebook/browser/find/notebookFindDecorations.ts [hunk 9]
```diff
@@ -189,51 +199,64 @@ export class NotebookFindDecorations implements IDisposable {
...
+			let codeCellFindScopes = findScopes.filter((c, i, ranges) => {
+				return ranges.indexOf(ranges.find(t => t.cell.cellGuid === c.cell.cellGuid && t.cell.cellType === 'code' && t.outputComponentIndex === -1)) === i;
 			});
+			if (codeCellFindScopes) {
+				this._editor.changeDecorations((accessor) => {
+					let findMatchesOptions = NotebookFindDecorations._FIND_MATCH_NO_OVERVIEW_DECORATION;
+					// filter code cell find matches
+					findMatches = findMatches.filter(m => m.range.cell.cellType === 'code' && m.range.outputComponentIndex === -1);
+					/* findMatches = findMatches.filter((c, i, matches) => {
+						return matches.indexOf(matches.find(t => t.range.cell.cellGuid === c.range.cell.cellGuid && t.range.cell.cellType === 'code' && !t.range.isCodeOutput)) === i;
```
Comment (barbaravaldez) [inline src/sql/workbench/contrib/notebook/browser/find/notebookFindDecorations.ts:221]: let's remove this commented code if we are not going to use it

**Review by lucyzhang929 — APPROVED** [review]
Tested and find in cell output works overall but I did notice one bug:
The highlight for the second html mime output never gets removed.
Repro steps:
1. Run SQL-WorldWideImport.ipynb
2. Search for "total"
3. Then search for something else
The original highlight for "time" does not get removed.

https://user-images.githubusercontent.com/23462877/146802225-e674b0fe-9aaf-4c11-a29e-1c9fb4056869.mov

**Comment by MaddyDev** [comment]
> Tested and find in cell output works overall but I did notice one bug: The highlight for the second html mime output never gets removed. Repro steps:
> 
> 1. Run SQL-WorldWideImport.ipynb
> 2. Search for "total"
> 3. Then search for something else
>    The original highlight for "time" does not get removed.
> 
>  Screen.Recording.2021-12-20.at.8.31.39.AM.mov

Thanks Lucy for testing this, I fixed it now. Appreciate your help!

**Review by barbaravaldez — APPROVED** [review]

**Pull request merged**